### PR TITLE
Fixes and changes for indentation

### PR DESCRIPTION
**Pull request by kopoli**
In #5 the following problematic points were discovered:

1. The enabled-by-default electric-indent-mode indented the current line wrong when pressing `RET`.
2. A feature request where the point could be retained when it is after the indentation.

### CODE REVIEW

**Comment by BrachystochroneSD** [comment]
1. This issue remains with current version:
electric_indent_problem

2. Also, when the cursor is not at the current indent, impossible to toggle between 0 and 1 indentation. 
Making the `robot-mode-retain-point-on-indent` useless. Since the code only allows the toggling when the cursor is at the curent-indent

3. You are allowing the toggling in "IF" block for example, which is not ideal (but we can maybe live with it).

File: robot-mode.el [hunk 1]
```diff
@@ -194,6 +201,10 @@ Used as `indent-line-function' of the mode."
 	    (string-width (buffer-substring-no-properties
 			   (line-beginning-position) (point)))))
 
+	 ;; Current point in line as width
+	 (point-in-line (string-width (buffer-substring-no-properties
+			   (line-beginning-position) (point))))
+
```
Comment (BrachystochroneSD) [inline robot-mode.el]: `(current-column)` gives you the same result

File: robot-mode.el [hunk 2]
```diff
@@ -194,6 +201,10 @@ Used as `indent-line-function' of the mode."
...
 	 ;; The non-indented contents of the current line
 	 current-line
```
Comment (BrachystochroneSD) [inline robot-mode.el]: Maybe set the current-line at this stage and not in the `current-indent` to be more clear.
Also note that there is a built-in function `(current-indentation)`, you could just use it instead of calculating it with `string-width`.

**Comment by BrachystochroneSD** [comment]
Another major problem with this implementation of indentation is that the completion (provided by eglot f.ex) can't be triggered with the usual "tab".

For example:
image
Note that this example has been made with the implementation proposed with #5. From your legacy code, the cursor go back to indentation and nothing is shown.

Mmmh, maybe it's because the cursor is moving, and the reason why it shouldn't move. (And maybe why no major-mode is doing it for "indent-for-tab-command", but I might be wrong).

Usually it's not problematic for fixed-indent language (like C) but not for something like python, gdscript or here, for robot. That was the crux of this indentation system. Allowing the cycling + the completion.

The way python mode solve this is by checking if the previous command was a `indent-for-tab-command` to trigger the indentation cycling. Allowing to trigger the completion if any between two tab. like:
```
Test Case 
     Keyw|     -> Tab Trigger completion if any, completing it to "Keyword"
```
```
Test Case
    Keyword|  -> Tab trigger now the cycling, since the previous command was a Tab as well and there is no more completion available
```
This solution has been chosen and adapted to robot in my #5 PR already, as you would imagine hehe.

**Comment by kopoli** [comment]
About the issues:

1. For this, setting the `electric-indent-inhibit` is probably the best
   option, as the first reindent can trigger this toggling.

2. Indeed, the toggling is enabled only at the current indent. What the
   `robot-mode-retain-point-on-indent` prevents is moving the point to current
   indent when it is after the current indent. I.e. it indents when point is
   anywhere in the line, but toggles the indent only when point is  in the
   current indent. 

3. Toggling also the indentation of control structures is something I wanted
   to retain. My goal for the toggling was that the developer knows best what
   the indentation should be, and the mode should make that easy.


About the completion and `TAB`. For some reason it appears to work for me at
least with `company-mode`. I'm able to apply the completion with `TAB` if
there is a single match. If there are multiple matches displayed in a list, I
can use the cursor keys and select with `RET`. This appears to work similarly
in other modes.

File: robot-mode.el (hunk 1 above)
Comment (kopoli) [reply robot-mode.el]: Thanks! I didn't know about this. Unfortunately Emacs documentation says that it was introduced around version 29.1. I would like to retain support for versions older than that.

Comment (BrachystochroneSD) [reply robot-mode.el]: I can confirm that it is available on emacs 27.1. It's a built-in function compiled in the C source. The emacs documentation can't really get the exact introduction of these kind of function it seems.
 
And in my opinion, emacs 27.1 is currently the version of the "oldstable" package of debian. So, we can consider it as outdated.

**Comment by BrachystochroneSD** [comment]
> Indeed, the toggling is enabled only at the current indent. What the
> robot-mode-retain-point-on-indent prevents is moving the point to current
> indent when it is after the current indent. I.e. it indents when point is
> anywhere in the line, but toggles the indent only when point is in the
> current indent.

Why not making the toggling possible from anywhere in the line?

> Toggling also the indentation of control structures is something I wanted
> to retain. My goal for the toggling was that the developer knows best what
> the indentation should be, and the mode should make that easy.

Why having an indentation system in the first place? If someone think it's best to toggle a keyword that is inside a "IF" statement to be at indent 0, he will get an error during execution. If someone don't want to use indentation system, he can just disabling it.

I will check with company mode.

**Comment by BrachystochroneSD** [comment]
I've checked with company.
The way company works is a bit different than the default completion system. 
Company-mode list the completions when the cursor is idling, then let you complete it by hitting RET, or (TAB if there is only one match as you mentioned correctly). 

The default completion system (and the one for counsel/ivy) is only called when needed by hitting "TAB" if the cursor is not moving (meaning, indentation is complete). 
This means that the default completion system is bypassed with the choice of moving the cursor back to indentation by default. 

I really feel like moving cursor back to indent when hitting tab is very counter-intuitive for people using emacs. This is how emacs works for a good amount of major mode. Plus, I don't see any IDE who move the cursor back to indentation when hitting tab either by the way. I don't have a complete knowledge of all major mode of course, by from my experience, this is a strange feature. And if you don't remember what was its usage, maybe it's because it was a legacy bug that you just get used to it?

**Comment by kopoli** [comment]
Could you give a configuration example for a completion system where things don't work with the toggling ?

From what I can tell, the default completion system for Emacs is: https://www.gnu.org/software/emacs/manual/html_node/elisp/Completion-in-Buffers.html
However, that uses `C-M-i` or `M-TAB` bindings for applying the completion (and therefore shouldn't clash with `TAB` usage).

**Comment by kopoli** [comment]
> Why not making the toggling possible from anywhere in the line?

You are correct. Setting the `electric-indent-inhibit` is a more complete solution than returning to the 0.6 behaviour fix. I removed that commit, and the toggling can be done anywhere from the line.

**Comment by BrachystochroneSD** [comment]
> Could you give a configuration example for a completion system where things don't work with the toggling ?
> 
> From what I can tell, the default completion system for Emacs is: https://www.gnu.org/software/emacs/manual/html_node/elisp/Completion-in-Buffers.html However, that uses `C-M-i` or `M-TAB` bindings for applying the completion (and therefore shouldn't clash with `TAB` usage).

You are right, in fact I think this is only a problem when using the option:
`(setq tab-always-indent 'complete)`

Which I use personally. I don't want to force a feature just for me. And I don't know how frequent is the usage of that option in the community. So I let you take the choice to consider this.
But I can confirm that this was a frustrating problem that I haven't encounter on other major-mode until now.
But the option to retain the cursor position kind of solve this.

I'll do some tests with this last version. Thanks for listening and considering these feedback by the way.

**Comment by BrachystochroneSD** [comment]
Hello,

Seems that now, the `C-a` to align the defun is now broken. It toggles lines as well. Maybe remove the `(indent-region beg end)` in the `robot-mode-align` to fix it? Not sure if it is needed for something else

**Comment by kopoli** [comment]
> Seems that now, the C-a to align the defun is now broken. It toggles lines as well. Maybe remove the (indent-region beg end) in the robot-mode-align to fix it? Not sure if it is needed for something else

Thanks for reporting this !

If the `robot-mode-basic-offset` is different than `robot-mode-argument-separator`, then the indentation would be off after alignment. I.e. the align function actually indents the contents according to `robot-mode-argument-separator` (if there is at least two spaces in the beginning of the line).

I opted for adding a `defvar` for controlling the toggling functionality (and disabling it while aligning). Hopefully that doesn't break anything else :sweat_smile: 

Do you happen to know if there is some kind of testing functionality for emacs that would simulate and compare buffer operations ? (It would be good to have some buffer level regression tests to catch breakage such as this.)

I mean the following kind of contents for a test:
1. Initialize the test with some buffer contents.
2. Run some Emacs operations/keypresses to the buffer.
3. New buffer contents with which the buffer under test should be equivalent.

**Comment by BrachystochroneSD** [comment]
Tested and it seems to be OK! Thanks. I'll continue to use this version if I found some other bugs. 

I've never did acceptance testing for emacs packages yet. 

I've done a little bit of research and I have found two packages [ERT](https://www.gnu.org/software/emacs/manual/html_mono/ert.html) (built-in) and [buttercup](https://github.com/jorgenschaefer/emacs-buttercup/tree/master). Pretty sure both could be used for what you've asked.

Seems pretty interesting.

File: robot-mode.el [hunk 3]
```diff
@@ -249,9 +252,10 @@ Used as `indent-line-function' of the mode."
 		 robot-mode-basic-offset)))
 
     ;; Toggle indentation if the line is already indented
-    (when (and (> indent 0)
-	       (= indent current-indent))
-      (setq indent 0))
+    (when robot-mode-indent-toggle 
+      (when (and (> indent 0)
+		 (= indent current-indent))
```
Comment (BrachystochroneSD) [inline robot-mode.el]:
```emacs-lisp
(when (and robot-mode-indent-toggle 
           (> indent 0)
	   (= indent current-indent))
```

File: robot-mode.el [hunk 4]
```diff
@@ -238,17 +252,20 @@ Used as `indent-line-function' of the mode."
...
-    (back-to-indentation)
+    (when (not robot-mode-retain-point-on-indent)
+      (back-to-indentation))
 
-    ;; Do the actual indenting if indentation changed
-    (when (not (= indent current-indent))
-      (delete-region (line-beginning-position)  (point))
-      (indent-to indent))))
+    ;; Save point if point retention is enabled and point is after indentation
+    (if (and (> point-in-line indent)
```
Comment (BrachystochroneSD) [inline robot-mode.el]: This is not correct, should be compared with the current indentation and not the calculated indentation. This cause the cursor at column < 4 to go back at indentation when toggling from indent=0 to indent=4 (given an indent offset to 4)

Also consider using `(current-column)` and `(current-indentation)`. This will save useless calculation in the `let` setter. (removing `current-indent` and `point-in-line`)

**Review by BrachystochroneSD — COMMENTED** [review]
I added some reviews and remarks.

Also the previously mentioned problem concerning the incapacity to complete with default completion system and `(setq tab-always-indent 'complete)` remains.

Do you plan to fix that?

File: robot-mode.el (hunk 3 above)
Comment (kopoli) [reply robot-mode.el]: Thanks! Updated.

File: robot-mode.el (hunk 4 above)
Comment (kopoli) [reply robot-mode.el]: > This is not correct, should be compared with the current indentation and not the calculated indentation. This cause the cursor at column < 4 to go back at indentation when toggling from indent=0 to indent=4 (given an indent offset to 4)

Isn't that how it's supposed to work?
IIUC e.g. in `emacs-lisp-mode` if the point is between `0` and indent offset, and you press `TAB`, then the point moves back to indentation (or rather forward to indentation). Or am I misunderstanding something ?

Added usage of `(current-column)` and `(current-indentation)`. Thanks!

I looked at the Emacs source for `current-column` when the documentation said that it was introduced/changed around version 29.1, but apparently that was the variable and not the function. The function has been present for ages.

**Comment by kopoli** [comment]
> Also the previously mentioned problem concerning the incapacity to complete with default completion system and (setq tab-always-indent 'complete) remains.

Could you give out a small elisp configuration that would enable the default completion system ?
I tried just setting the following:
```
(setq tab-always-indent 'complete)
```
But unfortunately it doesn't complete anything. I see that it uses `(completion-at-point)` inside (and that doesn't complete anything, which most likely is the problem here). Do I need to configure it somehow (e.g. with `hippie-expand` or somesuch) ?

**Comment by kopoli** [comment]
I was able to get some kind of completion using TAB with the following config:
```
(setq tab-always-indent 'complete)
(require 'dabbrev)
(setq-default completion-at-point-functions '(dabbrev-completion))
```

I added a commit that would prevent the toggling behaviour in favour of the TAB completion behaviour only when fully indented and point being after the indentation. Would this kind of logic work for your case ?

File: robot-mode.el (hunk 4 above)
Comment (BrachystochroneSD) [reply robot-mode.el]: This is absolutely correct, the cursor should follow the indentation when it is located within the current indentation when you press tab.

But here, it checks if it is within the indentation after setting the indentation. Here are some examples to illustrate the "problem":

1. In this case, it should follow the indentation (which is the case)

```robotframework
Test Case 1
| Keyword  
```
And become:
```robotframework
Test Case 1
    |Keyword  
```

2. In this second case, it shouldn't follow indentation and should save the excursion
```robotframework
Test Case 1
Key|word  
```
And should become:
```robotframework
Test Case 1
    Key|word  
```

Which is not the case currently, because it is compared with the `indent` (meaning, the calculated `indent`) and not the `current-indent`. 
From the last example:
  - `current-column = 3`
  - `current-indentation = 0` 
  - `indent (calculated-indent) = 4`
  - `current-column < indent` -> Follow indentation.

**Comment by BrachystochroneSD** [comment]
Almost there, because currently, you just disable toggling with the setup `tab-always-indent` is set to `complete`
Here is a solution that allow both:
```emacs-lisp
	       ;; Allow toggling when using tab-always-indent
               ;; by using tab twice instead of once.
	       (if (and robot-mode-retain-point-on-indent
			 (eq tab-always-indent 'complete))
		   (equal this-command last-command)
                 t))
```

File: robot-mode.el (hunk 4 above)
Comment (kopoli) [reply robot-mode.el]: Thanks for the explanation!
I was able to reproduce the situation, and apply the suggested fix :+1:

**Comment by kopoli** [comment]
> Almost there, because currently, you just disable toggling with the setup `tab-always-indent` is set to `complete` Here is a solution that allow both:
> 
> ```emacs-lisp
> 	       ;; Allow toggling when using tab-always-indent
>                ;; by using tab twice instead of once.
> 	       (if (and robot-mode-retain-point-on-indent
> 			 (eq tab-always-indent 'complete))
> 		   (equal this-command last-command)
>                  t))
> ```

Do I understand correctly that when the line is correctly indented, the first `TAB` should do the completion and the second `TAB` should do the toggling ?

I tested this with the above `dabbrev` configuration, and `TAB` sometimes seems to not call the `robot-mode-indent-line` (aka. `indent-line-function`) at all, and just skip to calling `completion-at-point`. Especially if I just move  the point in the same line (e.g. with `left-char` `right-char`).

I.e. I can see the first `TAB` goes through the `robot-mode-indent-line` function, and just disables the toggling (and subsequently calls `completion-at-point`). The next `TAB` just appears to call `completion-at-point` directly (skipping the possibility of toggling).

**Comment by kopoli** [comment]
> I.e. I can see the first TAB goes through the robot-mode-indent-line function, and just disables the toggling (and subsequently calls completion-at-point). The next TAB just appears to call completion-at-point directly (skipping the possibility of toggling).

Debugged this a bit further, and it seems that `completion-at-point` likely starts `completion-in-region-mode` via `completion-in-region`. That minor mode overrides the `TAB` to call directly the `completion-at-point` (skipping indentation altogether).

**Comment by BrachystochroneSD** [comment]
> Do I understand correctly that when the line is correctly indented, the first TAB should do the completion and the second TAB should do the toggling ?

With that proposition, toggling is only available if there is no more completion available. (or if you double tab quick enough so that the completion list is not yet ready)
This is the somewhat the same process used in `python-mode` to cycle between indentations

But indeed, this proposition (and the toggling in python-mode) doesn't seems to work with the setup with `dabbrev`. Seems to work with ivy tho. Meaning I was wrong when I mention the "default completion" system impact.

Here is a minimal configuration example:
```emacs-lisp
(setq tab-always-indent 'complete)
(require 'eglot)    ;; To have completion available in robot-mode
(require 'robot-mode)
(setq robot-mode-retain-point-on-indent t)
(with-eval-after-load 'eglot
  (defvar eglot-server-programs)
  (push (cons 'robot-mode '("robotframework_ls"))
        eglot-server-programs))  ;; add language server to the pool of eglot for completion
(require 'ivy)    ;; completion system
```

So, to sum up, that check with `last-command this-command` allow ivy-user (at least) with `tab-always-indent` set to `complete` to use toggling anywhere in the line (instead of just not being able to to it) without impacting other configuration.

Note that forcing push make it difficult for other people to work on your project and completely messing up the history shown in merge request, I would suggest to avoid it.

**Comment by kopoli** [comment]
> But indeed, this proposition (and the toggling in python-mode) doesn't seems to work with the setup with dabbrev. Seems to work with ivy tho. Meaning I was wrong when I mention the "default completion" system impact.

I think that my problem is that the `dabbrev-completion` function is completely unsuitable to use with `completion-at-point-functions`.
I found a `dabbrev` -based completion function in https://emacs.stackexchange.com/questions/30690/code-auto-completion-with-ivy and that seems to work properly (at least with minimal weird behaviour).

I added the `(equal this-command last-command)` check in the toggling. Does it now work as intended ?

> Note that forcing push make it difficult for other people to work on your project and completely messing up the history shown in merge request, I would suggest to avoid it.

I will force push changes in feature branches (under review) so that all fixups end up in their proper commits. The reason is that then the commit history will be clear with sensible commits when this is pushed to master. I will do a fast-forward merge or with an empty merge-commit when this ultimately goes in.

An alternative would be to push all fixups at the end of the branch and finally do a squash-merge to master. This would be confusing during review as there are multiple commits going back and forth between proposed implementations, and it would be confusing in master where numerous unrelated changes would be lumped up in a single commit.


Finally, I added a change to properly indent situations when IF/ELSE/etc. word is embedded in a keyword. This is inspired from #10. (I didn't include it in #11 as that has syntax highlight fixes, and this one has indentation fixes.)
